substrate/client/transaction-pool/src/graph/base_pool.rs

```diff
@@ -322,22 +354,36 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> BasePool<Hash,
 
 			// import this transaction
 			let current_hash = tx.transaction.hash.clone();
+			let current_tx = tx.transaction.clone();
 			match self.ready.import(tx) {
 				Ok(mut replaced) => {
 					if !first {
-						promoted.push(current_hash);
+						promoted.push(current_hash.clone());
 					}
+					// If there were conflicting future transactions promoted, removed them from
+					// promoted set.
+					promoted.retain(|hash| replaced.iter().all(|tx| *hash != tx.hash));
 					// The transactions were removed from the ready pool. We might attempt to
 					// re-import them.
 					removed.append(&mut replaced);
 				},
+				Err(e @ error::Error::TooLowPriority { .. }) =>
```

Comment:
nit: `PriorityTooLow`

Reply:
Sorry, I am not keen to rename long existing enums in this PR.